src/main/java/com/minecolonies/coremod/colony/Structures.java

```diff
@@ -60,200 +109,845 @@ public static void init()
+     * check that we can store the schematic.
+     * According to the total number of schematic allowed on the server
+     *
+     * @return true if we can store more schematics
+     */
+    private static boolean canStoreNewSchematic()
+    {
+        if (MineColonies.isClient())
+        {
+            return true;
+        }
+        if (!Configurations.allowPlayerSchematics)
+        {
+            return false;
+        }
+
+        final Set<String> md5Set = getCachedMD5s();
+        Log.getLogger().info("Server has " + md5Set.size() + " cached schematics");
+        if (md5Set.size() < Configurations.maxCachedSchematics)
+        {
+            return true;
+        }
+
+
+        int countInUseStructures = 0;
+        for (final Colony c : ColonyManager.getColonies())
+        {
+            Log.getLogger().info("Looking a workorder in Colony " + c.getName());
+            for (final AbstractWorkOrder workOrder : c.getWorkManager().getWorkOrders().values())
+            {
+                if (workOrder instanceof WorkOrderBuildDecoration)
+                {
+                    final String schematicName = ((WorkOrderBuildDecoration) workOrder).getStructureName();
+                    Log.getLogger().info("Looking a workorder with structure " + schematicName);
+                    if (md5Set.contains(schematicName))
+                    {
+                        Log.getLogger().info("Colony " + c.getName() + " use cached schematic " + schematicName);
+                        md5Set.remove(schematicName);
+                        countInUseStructures++;
+                    }
+                }
+            }
+        }
+        Log.getLogger().info("Server has " + countInUseStructures + " used cached schematics ");
+        Log.getLogger().info("Server has " + md5Set.size() + " not used cached schematics ");
+        //md5Set containd only the unused one
+        Iterator<String> iterator = md5Set.iterator();
+        while (iterator.hasNext() && md5Set.size() + countInUseStructures >= Configurations.maxCachedSchematics)
+        {
+            final StructureName sn = new StructureName(iterator.next());
+            if (deleteCachedStructure(sn))
+            {
+                Log.getLogger().info("Removing cached schematic " + sn);
+                iterator.remove();
+            }
+        }
+
+        return md5Set.size() + countInUseStructures < Configurations.maxCachedSchematics;
+    }
+
+    /**
+     * Save a schematic in the cache.
+     * This method is valid on the client and server
+     * The schematic will be save under the cache directory using is md5 hash as a name.
+     *
+     * @param bytes representing the schematic.
+     * @return True is the schematic have been saved successfully.
+     */
+    public static boolean handleSaveSchematicMessage(final byte[] bytes)
+    {
+        final File schematicsFolder = Structure.getCachedSchematicsFolder();
+        final String md5 = Structure.calculateMD5(bytes);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Move the declaration of "md5" closer to the code that uses it. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1941)